### PR TITLE
feat: LT-1142 - Goods docs

### PR DESCRIPTION
**Pull request by cstobart**

### CODE REVIEW

File: documents/tasks.py [hunk 1]
```diff
@@ -9,4 +9,5 @@ def prepare_document(document_id):
     """
     from documents.models import Document
     doc = Document.objects.get(id=document_id)
+    print('found document', doc)
```
Comment (5pence) [inline documents/tasks.py]: remove print statement

File: documents/views.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+class DocumentDetail(APIView):
+    authentication_classes = (GovAuthentication,)
+
+    def get(self, request, pk):
+        """
+        Returns a list of documents on the specified good
+        """
+        print(pk)
+        print(Document.objects.all())
```
Comment (5pence) [inline documents/views.py]: remove these print statements

File: documents/views.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        document = Document.objects.get(id=pk)
+        serializer = DocumentViewSerializer(document)
+        return JsonResponse({'document': serializer.data})
```
Comment (5pence) [inline documents/views.py]: add new line at end of file for pep8

File: drafts/views/draft_goods.py [hunk 4]
```diff
@@ -48,7 +49,11 @@ def post(self, request, pk):
 
         organisation = get_organisation_by_user(request.user)
         get_draft_with_organisation(pk, organisation)
-        get_good_with_organisation(data.get('good'), organisation)
+        good = get_good_with_organisation(data.get('good'), organisation)
+
+        if len(GoodDocument.objects.filter(good=good)) == 0:
```
Comment (5pence) [inline drafts/views/draft_goods.py]: if len(GoodDocument.objects.filter(good=good)) == 0: unnecessary len used - inefficient

File: goods/migrations/0003_gooddocument.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+    operations = [
+        migrations.CreateModel(
+            name='GoodDocument',
+            fields=[
+                ('document_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, serialize=False, to='documents.Document')),
+                ('description', models.TextField(blank=True, default=None, max_length=280, null=True)),
+                ('good', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='goods.Good')),
+                ('organisation', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='organisations.Organisation')),
+                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='users.ExporterUser')),
```
Comment (5pence) [inline goods/migrations/0003_gooddocument.py]: not sure you want cascading deletion on user being deleted - has an organisation could have a user that leaves but wants to keep docs against a license

File: goods/serializers.py [hunk 6]
```diff
@@ -97,3 +105,49 @@ def update(self, instance, validated_data):
...
+    def create(self, validated_data):
+        good_document = super(GoodDocumentCreateSerializer, self).create(validated_data)
+        good_document.save()
+
+        if BACKGROUND_TASK_ENABLED:
+            prepare_document(str(good_document.id))
+        else:
+            try:
+                prepare_document.now(str(good_document.id))
+            except Exception:
```
Comment (5pence) [inline goods/serializers.py]: please declare what type of exceptions

File: goods/views.py [hunk 7]
```diff
@@ -114,6 +99,111 @@ def delete(self, request, pk):
...
+        good_documents = GoodDocument.objects.filter(good=good).order_by('-created_at')
+        serializer = GoodDocumentViewSerializer(good_documents, many=True)
+
+        return JsonResponse({'documents': serializer.data})
+
+    @swagger_auto_schema(
+        request_body=GoodDocumentCreateSerializer,
+        responses={
+            400: 'JSON parse error'
+        })
```
Comment (5pence) [inline goods/views.py]: add blank line underneath

File: goods/migrations/0003_gooddocument.py (hunk 5 above)
Comment (cstobart) [reply goods/migrations/0003_gooddocument.py]: users don't ever get deleted, but I get your point

File: goods/tests/tests_goods_documents.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+        response = self.client.get(self.url, **self.exporter_headers)
+        response_data = response.json()
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response_data['documents']), 2)
+
+
+    # @tag('slow')
+    # def test_can_remove_document_from_unsubmitted_good(self):
```
Comment (jfaracik) [inline goods/tests/tests_goods_documents.py]: Could you write an explanation?

File: goods/urls.py [hunk 9]
```diff
@@ -6,5 +6,7 @@
 
 urlpatterns = [
     path('', views.GoodList.as_view(), name='goods'),
```
Comment (jfaracik) [inline goods/urls.py]: Add # ex:

**Comment by efahy** [comment]
Please investigate the possibility of adding Jim's Path to `Pipfile`

**Comment by efahy** [comment]
Please make the error presented when you don't attach a doc to a good more informative.  "An Error Occurred" is not telling the user a whole lot

**Comment by efahy** [comment]
The unsure comment field should be mandatory when adding an unsure good

File: goods/models.py [hunk 10]
```diff
@@ -16,3 +18,10 @@ class Good(models.Model):
     part_number = models.TextField(default=None, blank=True, null=True)
     organisation = models.ForeignKey(Organisation, on_delete=models.CASCADE, default=None)
     status = models.CharField(choices=GoodStatus.choices, default=GoodStatus.DRAFT, max_length=20)
+
+
+class GoodDocument(Document):
+    good = models.ForeignKey(Good, on_delete=models.CASCADE)
+    user = models.ForeignKey(ExporterUser, on_delete=models.CASCADE)
+    organisation = models.ForeignKey(Organisation, on_delete=models.CASCADE)
```
Comment (efahy) [inline goods/models.py]: When users and orgs are deleted, please leave the documents

File: goods/serializers.py [hunk 11]
```diff
@@ -97,3 +105,50 @@ def update(self, instance, validated_data):
...
+            # pylint: disable=W0703
+            try:
+                prepare_document.now(str(good_document.id))
+            except Exception:
+                raise serializers.ValidationError({'errors': {'document': 'Failed to upload'}})
+
+        return good_document
+
+
+class GoodDocumentViewSerializer(serializers.ModelSerializer):
```
Comment (efahy) [inline goods/serializers.py]: As discussed, the API should return a pre-canned download link, not a self assembly kit.  Please reflect this anywhere else that the kit is being returned

File: goods/views.py [hunk 12]
```diff
@@ -114,6 +99,111 @@ def delete(self, request, pk):
...
+    def post(self, request, pk):
+        """
+        Adds a document to the specified good
+        """
+        good = get_good(pk)
+        good_id = str(good.id)
+        data = request.data
+        organisation = get_organisation_by_user(request.user)
+
+        if good.organisation != organisation:
```
Comment (efahy) [inline goods/views.py]: In the event of any errors, we need to delete the file from S3 and not leave it in there orphaned (but do it on a best endeavour basis) (I realise there are complications, just make it work Chris)

File: goods/serializers.py (hunk 11 above)
Comment (cstobart) [reply goods/serializers.py]: We don't actually construct a download link, its a pre built s3 call which takes the s3_key as a method parameter so will not be changing this

**Pull request merged**